### PR TITLE
Enh: Toast widget

### PR DESCRIPTION
**Pull request by Mister-42**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Fixed issues  | 

Thanks @simialbi

### CODE REVIEW

File: tests/ToastTest.php [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+</div></div>
+HTML;
+
+        $this->assertEqualsWithoutLE($expected, $out);
+    }
+
+    public function testDateTimeOptions()
+    {
+        Toast::counter(0);
+        $out = Toast::widget()
```
Comment (terabytesoftw) [inline tests/ToastTest.php:74]: you can name it html so all other tests will be the same.

File: tests/ToastTest.php [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    }
+
+    /**
+     * @depends testBodyOptions
+     */
+    public function testContainerOptions()
+    {
+        Toast::counter(0);
+
+        ob_start();
```
Comment (terabytesoftw) [inline tests/ToastTest.php:46]: don't use ob_start, use string output.

**Review by terabytesoftw — APPROVED** [review]

File: src/Toast.php [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+ * For example,
+ *
+ * ```php
+  * Toast::widget()
+ *     ->title('Hello world!')
+ *     ->dateTime('a minute ago')
+ *     ->body('Say hello...')
+ * ])
+ * ->begin();
+ * ```
```
Comment (terabytesoftw) [inline src/Toast.php:25]: I think the example is wrong, can you verify it thanks?

 ```php
 Toast::widget()
     ->title('Hello world!')
     ->dateTime('a minute ago')
     ->body('Say hello...')
     ->begin();
 ```

File: src/Toast.php [hunk 4]
```diff
@@ -0,0 +1,402 @@
...
+ *
+ * The following example will show the content enclosed between the [[begin()]]
+ * and [[end()]] calls within the toast box:
+ *
+ * ```php
+ * Toast::widget()
+ *     ->title('Hello world!')
+ *     ->dateTime('a minute ago')
+ * ])
```
Comment (terabytesoftw) [inline src/Toast.php:34]: Same here.

File: src/Toast.php [hunk 5]
```diff
@@ -0,0 +1,402 @@
...
+ *
+ * @see https://getbootstrap.com/docs/components/toasts/
+ *
+ * @author Simon Karlen <simi.albi@outlook.com>
+ */
+final class Toast extends Widget
+{
+    /**
+     * @var string the body content in the alert component. Note that anything between
+     * the [[begin()]] and [[end()]] calls of the Toast widget will also be treated
```
Comment (terabytesoftw) [inline src/Toast.php:50]: * the {@see begin()} and {@see end()} calls of the Toast widget will also be treated

File: src/Toast.php [hunk 6]
```diff
@@ -0,0 +1,402 @@
...
+     * @var array the options for rendering the close button tag.
+     * The close button is displayed in the header of the toast. Clicking on the button will hide the toast.
+     *
+     * The following special options are supported:
+     *
+     * - tag: string, the tag name of the button. Defaults to 'button'.
+     * - label: string, the label of the button. Defaults to '&times;'.
+     *
+     * The rest of the options will be rendered as the HTML attributes of the button tag.
+     * Please refer to the [Toast documentation](https://getbootstrap.com/docs/4.5/components/toasts/)
```
Comment (terabytesoftw) [inline src/Toast.php:78]: upadate link https://getbootstrap.com/docs/5.0/components/toasts

File: src/Toast.php [hunk 7]
```diff
@@ -0,0 +1,402 @@
...
+     */
+    private array $closeButton = [];
+    /**
+     * @var array additional title options
+     *
+     * The following special options are supported:
+     *
+     * - tag: string, the tag name of the button. Defaults to 'strong'.
+     *
+     * @see \yii\helpers\Html::renderTagAttributes() for details on how attributes are being rendered.
```
Comment (terabytesoftw) [inline src/Toast.php:89]: update {@see Yiisoft\Html\Html::renderTagAttributes()}

File: src/Toast.php [hunk 8]
```diff
@@ -0,0 +1,402 @@
...
+     */
+    public array $titleOptions = [];
+    /**
+     * @var array additional date time part options
+     *
+     * The following special options are supported:
+     *
+     * - tag: string, the tag name of the button. Defaults to 'small'.
+     *
+     * @see \yii\helpers\Html::renderTagAttributes() for details on how attributes are being rendered.
```
Comment (terabytesoftw) [inline src/Toast.php:99]: Same here.

File: src/Toast.php [hunk 9]
```diff
@@ -0,0 +1,402 @@
...
+     */
+    private array $dateTimeOptions = [];
+    /**
+     * @var array additional header options
+     *
+     * @see \yii\helpers\Html::renderTagAttributes() for details on how attributes are being rendered.
```
Comment (terabytesoftw) [inline src/Toast.php:105]: Same Here.

File: src/Toast.php [hunk 10]
```diff
@@ -0,0 +1,402 @@
...
+     */
+    private array $bodyOptions = [];
+    /**
+     * @var array options
+     */
+    private array $options = [];
+    /**
+     * @var array footer options
+     *
+     * @see \yii\helpers\Html::renderTagAttributes() for details on how attributes are being rendered.
```
Comment (terabytesoftw) [inline src/Toast.php:121]: Same here.

**Comment by terabytesoftw** [comment]
Update classes with https://getbootstrap.com/docs/5.0/components/toasts/ ?

**Comment by terabytesoftw** [comment]
👍

File: src/Toast.php [hunk 11]
```diff
@@ -0,0 +1,398 @@
...
+    /**
+     * @var array additional title options
+     *
+     * The following special options are supported:
+     *
+     * - tag: string, the tag name of the button. Defaults to 'strong'.
+     *
+     * @see Html::renderTagAttributes() for details on how attributes are being rendered.
+     */
+    public array $titleOptions = [];
```
Comment (samdark) [inline src/Toast.php:87]: Should it be public? I don't think so.

Comment (samdark) [reply src/Toast.php:87]: https://github.com/yiisoft/yii-bootstrap5/commit/d0abc296dd4c5bdce784d9469d288f6969ce97b8

Comment (Mister-42) [reply src/Toast.php:87]: I'm not sure. Basically this is just an edited version of https://github.com/yiisoft/yii-bootstrap5/blob/master/src/Modal.php

Comment (samdark) [reply src/Toast.php:87]: There it wasn't public. Also I've made some fixes: https://github.com/yiisoft/yii-bootstrap5/commit/a2749520a40cce86a1a98a64867acdea7b829d83 Likely protected methods aren't correct in Modal as well.

**Pull request merged**